### ISSUE TITLE
✔️ 🚀 Feature Request: Column gap alignment issue

### ISSUE THREAD

**Issue by andreaswilthil** [elementor/elementor#452]
Thanks for a great plugin. We are considering moving from Visual Composer to Elementor in our projects, so we are currently trying your plugin .

We are having issues with the content of columns having a padding so the text is not aligned to other elements on the page, like header logo, post-title, footer-columns, etc.

Here is an example:
![issue](https://cloud.githubusercontent.com/assets/11750533/17982172/05a3d216-6b07-11e6-8a35-174403e9d7a0.jpg)

In my opinion, Visual Composer does this better, as the column content is aligned to elements outside the content-area.

**Comment by WPDevHQ** [comment]
In my tests I've found the main reason for this to be the default section width which is set at 1140px. The aligment is more noticeable if the theme's default content width is greater than that of the plugin.

I've been resolving this by increasing the section width to match that of my themes which tend to be set at 1280px as a default width.

**Comment by andreaswilthil** [comment]
In my case, the content width is not the issue. In my opinion the inner content of the first column in a row column should be aligned to the container that holds the Elementor elements. It could be solved by the parent element/row having a negative left-margin of the same value as the column padding.

**Comment by andreaswilthil** [comment]
The following css ( sass ) fixes the issue for me:

.elementor-container { 
			
			&.elementor-column-gap-default {
				margin-left:-10px !important; 
				margin-right:-10px !important;
			}
		
			&.elementor-column-gap-narrow {
				margin-left:-5px !important; 
				margin-right:-5px !important;
			}
			
			&.elementor-column-gap-extended {
				margin-left:-15px !important; 
				margin-right:-15px !important;
			}	
			
			&.elementor-column-gap-wide {
				margin-left:-20px !important; 
				margin-right:-20px !important;
			}
			
			&.elementor-column-gap-wider {
				margin-left:-30px !important; 
				margin-right:-30px !important;
			}
		}

**Comment by arielk** [comment]
If you are using in one column you can configure it with NO GAP.

We're going to add a global control for content width, I hope it will solve this problem

**Comment by daviedR** [comment]
@arielk the NO GAP solution doesn't work on my perspective. The actual problem is when you have columns gap. Here's the explanation:

The reason I consider this as an **issue** is that every time a user changes the gap size, user needs to adjust the content width option to get the same alignment as other sections.

For example:
* I want to have 1140px content width, means my whole text inside the section would be 1140px.
* With the default gap size, I have to set the Elementor's content width value to **1160px**, right? Because Elementor took 10px on left and right for the gap, so I added 20px on the content width setting. Make sense?
* And then when I change columns gap to "Extended", which gave the section 15px padding on the left and right, I **HAVE TO** adjust the Elementor's content width to **1170px**.
* If I set the columns gap to "Wider", I have to adjust the Elementor's content width to **1200px**.
* and so on..

I think this is quite a problem for end users. Every time they changed the columns gap, they have to do a little math to adjust the content width accordingly.

It would be a lot easier if whatever the columns gap value is, users will always see their text width matching the content width setting (in the example case is 1140px). They don't have to adjust the content width again. Don't you think so?

We can turn this into something better with this CSS:

```
.elementor-column-gap-default .elementor-row {
	width: calc(100% + 20px);
	margin-left: -10px;
	margin-right: -10px;
}
.elementor-column-gap-narrow .elementor-row {
	width: calc(100% + 10px);
	margin-left: -5px;
	margin-right: -5px;
}
.elementor-column-gap-extended .elementor-row {
	width: calc(100% + 30px);
	margin-left: -15px;
	margin-right: -15px;
}
.elementor-column-gap-wide .elementor-row {
	width: calc(100% + 40px);
	margin-left: -20px;
	margin-right: -20px;
}
.elementor-column-gap-wider .elementor-row {
	width: calc(100% + 60px);
	margin-left: -30px;
	margin-right: -30px;
}
```

Let me know what you think.

**Comment by enkota** [comment]
The issue I see is when you start building slightly more complex layouts. The gutters aren't all the same width. If we have 3 columns with default spacing between them, the middle column will always look like double the spacing as it's taking the left and right columns gutter.

<img width="1680" alt="screen shot 2016-10-13 at 16 34 49" src="https://cloud.githubusercontent.com/assets/13979799/19355834/b3264d4c-9163-11e6-94c2-92e4e6b4ad14.png">

**Comment by enkota** [comment]
@arielk any update with this? I'm having a hard time trying to build up some more advanced grid designs. For instance, in the design below, I am looking to make a grid using columns and background images. The problem begins with using nested columns in certain areas as the gutters won't work equally. I've attached the JSON template to show you how far I can get :) 

[elementor-390-2016-11-03.json.zip](https://github.com/pojome/elementor/files/569855/elementor-390-2016-11-03.json.zip)



![elementor](https://cloud.githubusercontent.com/assets/13979799/19980455/f8dd98ac-a1f5-11e6-903a-73fd0d44dc6f.jpg)

**Comment by fuzzyness** [comment]
I'm having the same problem as @daviedR. 

I have a section with a full width colour background. This section contains a nested 1 column widget and a 3 column widget (both inside the same parent section with the coloured background). 

The vertical left and right edges of these 2 sets of columns line up perfectly with each other. But... they don't line up with subsequent columns further down the page that aren't nested.

The CSS @daviedR posted works perfectly.

**Comment by daviedR** [comment]
Hi @enkota,

For your first image, if you want to have same gutter (gap) space, you can use left-right padding to your columns widget.

As for the second image, did you apply my CSS hack already?

**Comment by pmmueller** [comment]
@arielk Any update or further comment on this issue? Or is it a "feature" and gonna stay this way? 

I love working with Elementor, but this padding on the left and right are quite annoying and makes proper alignment of columns pretty difficult. Your suggestion to use "no gap" isn't really an option as in many cases I DO want a gap, but only *between* the columns. 

The following picture shows this problem on a simple page. All sections have default column gaps, but for some reason the blue section with the shape divider does not have the 10px left and right. 

![image](https://cloud.githubusercontent.com/assets/1928205/25072725/212107d8-22d7-11e7-8b14-8539b186b6c5.png)

The CSS hack from @daviedR does seem to work, but it feels weird to have/buy a design-oriented tool such as Elementor and then having to use a manual CSS hack. 

Question: 
If you don't want to change the default behavior it might be a help to add an option to remove the default padding at the very beginning and at the very end?

**Comment by newoceans** [comment]
Just wondering, anything new here?

I added this code, that seems to work for at least my own site right now (though @daviedR code might be cleaner, I don't know - since I'm not a programmer):

```
.elementor-column-gap-default>.elementor-row>.elementor-column>.elementor-element-populated {
    padding:0;
}
```

But where's the global setting @arielk mentioned at August 30th 2016?

**Comment by sethstha** [comment]
Why not make it like bootstrap? A row shouldn't have any padding while column can have it. This way my header and content will have an equal width and there will be no padding.

**Comment by ge022** [comment]
Why does .elementor-element-populated still have hardcoded padding?

**Comment by black-eye** [comment]
Also interested, whether there's something new on this issue.

**Comment by stephengardner** [comment]
Dear Elementor...
Creating a first section on a page would hopefully have that section touching the header.  Instead it's actually gapped by 10 pixels.

Is this not a huge problem?

You can't create that standard jumbotron look of having a large header image unless you use negative margins all over the place.

Seriously?

I'm adding custom CSS.  I just bought this plugin but I'm kinda regretting it now.

**Comment by hybridvision** [comment]
You have a couple of options:

1. Select "no gap" for this specific section in the Column Gap settings.
2. Add a custom CSS rule to your theme stylesheet to globally override the default gap (assuming you don't want gaps anywhere by default).

Neither option is too difficult and while it would be good if this was a little easier to change, I don't think there is a perfect setting for everyone. I'm sure plenty of people would complain if there was no gap by default...

**Comment by daviedR** [comment]
It's already more than a year, but still no real action from Elementor team. :(

**Comment by nicomollet** [comment]
Subscribring. Also interested in a solution for this.

**Comment by drank** [comment]
@daviedR, it is very unfortunate that this is just another issue that is being ignored by the Elementor Team, same as https://github.com/pojome/elementor/issues/78. Just another proof that fixing deep core issues is not their priority.

Do you also think that the energy and efforts spent on creating this video - https://www.youtube.com/watch?v=ZSohtminKQc - could have been utilized in a much more efficient way? :)

Sigh...

**Comment by arielk** [comment]
@drank Are you real? 

Have you checked all the issues that have already been closed? https://github.com/pojome/elementor/issues?q=is%3Aissue+is%3Aclosed

What you consider to be a problem is considered by another user as a feature, like this issue for example. (Try to add background to the column with text without padding).

We are constantly trying to improve the code, UI & UX, to be a better editor for web designers. There are things that take time to understand how to deal with them in a good manner.

Thanks

**Comment by drank** [comment]
@arielk, what's with the personal attacks?

I am expressing my personal opinion about how Elementor Team addresses 1.5-years old requests and you can't stand the critique so you need to get personal? Nice...

Anyway, the mere number of closed issues is nothing but "quantity over quality" and proves nothing in the discussed topic.

In fact, you're proving my point.

PS: I have nothing personal towards you, I don't know you and even if I did I don't think that I would have anyway. So please, try to read beyond the frustration and make some conclusions.

**Comment by daviedR** [comment]
@arielk, @drank, just chill guys... 

> "Try to add background to the column with text without padding"

That has nothing to do with this topic. But I just notice that it's another issue of your wrong concept of "column". I will create another issue topic about that later.

Back to the topic, let me elaborate this over again, I would make it really slowly this time.

- I am using a theme with 1140px content width.
- Elementor has global `Content Width` settings on Elementor > Settings. This is good feature. Let's say we set it to 1140px, because the theme's content width is 1140px too. Make senses, right?
- We create a page with Elementor now.
- I create a section with 2 columns, the default `Columns Gap` value is set to `Default` which has 10px padding on each column (right?).
- I put elements inside the columns.
- Boom! Now you see that the elements inside the columns are not aligned the same as theme's content width. Why? Because of those 10px.

A visual for you:

![image](https://user-images.githubusercontent.com/1287713/36849675-37cbeb40-1d97-11e8-84da-075b4479a812.png)

**"Well, we can adjust the Content Width to 1160px to accommodate that 10px gap on each side"**

Yes, you can. But don't you think it's really not user friendly? People have to adjust the content width every time they have sections with different gap. Besides, do all people have to remember that your `Default` columns gap is 10px padding or `Wide` columns gap is 20px padding, so they can do the math to readjust the content width value?

Why not make it like this:

![image](https://user-images.githubusercontent.com/1287713/36849697-444c263c-1d97-11e8-976d-0b88c47dd81e.png)

**Columns gap** should be the gap between each column. The edge of the 1st column and last column should be automatically aligned with the `content width` container. This is much more user friendly.

What do you think?

**Comment by vanjadin** [comment]
+1

Thanks @daviedR, 
I think you made the point pretty clear in your last post. 

Thx also for the CSS Hack - figuring out solutions for the padding on the outer columns already cost me quite some time when building custom layouts... not forgetting ensuing issues with the padding behaviour on smaller screens.

I think Elementor is a great plugin but consider this issue as too cumbersome for designing good looking websites.

**Comment by gschoppe** [comment]
+1

This could be accomplished by switching to a bootstrap 3 style positive padding, negative margin grid, or a flex-based space-between layout. A huge side benefit of either route would be to open the door to support for arbitrarily-deep column nesting, which would massively improve responsive options for complex grid layouts.

**Comment by ajlcomputers** [comment]
Just wanted to add my +1 to this.

I've actually just created a bug/enhancement request for this issue as I didn't find this post until afterwards. I guess it can be merged in here if required. It troubles me that it's heading towards 2 years and still this hasn't been resolved.

Currently if I want a 4 column layout with equal 20px spacing I have to create a 2 column layout and put 10px margin on the right of the left column and then 10px margin on the left of the right column. I then have to put 2 more columns in the left column and add the same margins to that and then 2 more columns in the right column with the same margins again so that I have 20px gap between each column and they're all the same size with no spacing on the far left or far right.

Having to do all that rather than someone just fixing the column gap option within Elementor is a massive pain.

**Comment by whatsnewsisyphus** [comment]
+1 for a simple solution
By default: Have a negative margin on the larger container to match column margins, so that the margin is between columns like every other grid system, not between the column and the overall container.
Add option: Have a page wide setting for a margin on the container.

The current version really doesn't make any sense as it throws all columns and content out of alignment with the rest of the site.

**Comment by daviedR** [comment]
Just a heads up, I created a plugin to enable columns alignment fix on Elementor sections and columns widget. You can optionally enable the fix on each section.

https://wordpress.org/plugins/columns-alignment-fix-for-elementor/

This will help you to have consistent content width without having to manually readjust content width value on each section when using columns gap.

**Comment by Hanleymade** [comment]
@daviedR Thanks for the plugin! 
I'm curious why your plugin solution can't be coded into the Elementor Builder?

**Comment by kelseyclubb** [comment]
@daviedR I truly appreciate your work on the plugin!

Unfortunately, it is not fully responsive.

It makes the screen wider than the mobile device. This causes horizontal scroll and is not mobile-friendly.

Not sure if this would be an easy fix, but I won't be able to utilize the plugin unless the mobile-friendliness is addressed. I assume others would care about this too.

**Comment by daviedR** [comment]
Hi @kelseyclubb ,

Can you create an issue here: https://github.com/sukiwp/columns-alignment-fix-for-elementor/issues/new

Please provide a screenshot about the "horizontal scroll" issue.

**Comment by RibeyeDesign** [comment]
@daviedR  Massive thank you for the plugin! Many hours struggling to figure this one out, led me to this thread, your plugin resolved it instantly.

**Comment by reachdigitalaus** [comment]
Hi @arielk 

Was this ever fixed, i tried Columns Gap " no gap " and I still get the 10px all around?

Thank you, and I look forward to your response.

**Comment by swinggraphics** [comment]
I used the plugin by @daviedR and customized it for my theme to make the default, narrow, extended, etc gaps more appropriate for my design. This adds margins for sections to prevent the content from butting up against the edge of window. I added this to my theme's functions.php:

```
function fpbc_alignment_fix_css() {
	if( class_exists( 'Columns_Alignment_Fix_For_Elementor' ) ) {
		// Use our own styles in place of those from plugin
		$caffe_instance = Columns_Alignment_Fix_For_Elementor::instance();
		remove_action( 'elementor/frontend/after_enqueue_styles', array( $caffe_instance, 'add_css' ) );
	}
	$container = get_option( 'elementor_stretched_section_container', 'body' );
	if( empty( $container ) ) {
		$container = 'body';
	}
	$container_width = get_option( 'elementor_container_width', 1140 );
	if( empty( $container_width ) ) {
		$container_width = 1140;
	}
	$space = get_option( 'elementor_space_between_widgets', 20 );
	if( empty( $space ) ) {
		$space = 20;
	}
	$css = $container . ' { overflow-x: hidden; }';
	$css .= '@media (max-width: ' . ( $container_width + $space * 2 ) . 'px) { .elementor-section.elementor-section-boxed > .elementor-container { max-width: calc(100% - ' . $space * 2 . 'px); } }';
	wp_add_inline_style( 'elementor-global', $css );
}
add_action( 'elementor/frontend/after_enqueue_styles', 'fpbc_alignment_fix_css', 1 );
```

And this is in style.css:

```
.elementor-column-gap-default > .elementor-row > .elementor-column > .elementor-element-populated {
	padding: 15px;
}
.elementor-column-gap-narrow > .elementor-row > .elementor-column > .elementor-element-populated {
	padding: 10px;
}
.elementor-column-gap-extended > .elementor-row > .elementor-column > .elementor-element-populated {
	padding: 20px;
}
.elementor-column-gap-wide > .elementor-row > .elementor-column > .elementor-element-populated {
	padding: 25px;
}
.elementor-column-gap-wider > .elementor-row > .elementor-column > .elementor-element-populated {
	padding: 30px;
}
.elementor-section.elementor-columns-alignment-fix-enabled > .elementor-column-gap-default > .elementor-row {
	width: calc(100% + 30px); margin-left: -15px; margin-right: -15px;
}
.elementor-section.elementor-columns-alignment-fix-enabled > .elementor-column-gap-narrow > .elementor-row {
	width: calc(100% + 20px); margin-left: -10px; margin-right: -10px;
}
.elementor-section.elementor-columns-alignment-fix-enabled > .elementor-column-gap-extended > .elementor-row {
	width: calc(100% + 40px); margin-left: -20px; margin-right: -20px;
}
.elementor-section.elementor-columns-alignment-fix-enabled > .elementor-column-gap-wide > .elementor-row {
	width: calc(100% + 50px); margin-left: -25px; margin-right: -25px;
}
.elementor-section.elementor-columns-alignment-fix-enabled > .elementor-column-gap-wider > .elementor-row {
	width: calc(100% + 60px); margin-left: -30px; margin-right: -30px;
}
```

Now if Elementor would just let us set padding/margins for some sides without having to set them all!

**Comment by seriusokhatsky** [comment]
still problem in 2020. And the fix is not appliable in Elementor 3.0. 
Elementor devs, are you working on some solution to this? After 3.0 update, all users that used this fix will have to manually go through thousands of columns and reset paddings here and there. Why not add some option in Elementor core?

**Comment by daviedR** [comment]
Elementor 3.0 removes the `elementor-row` element in section's HTML markup. This means the custom previously proposed CSS in the ticket is no longer valid.

**Comment by seriusokhatsky** [comment]
Hi, @daviedR 

So there is no solution to this in Elementor 3.0? Should we manually reset all paddings now or you will work on some general solution to this? It would be great to have some options for this. Otherwise, columns gap option for sections because useless since you have to reset left and right paddings every time.

**Comment by daviedR** [comment]
I am afraid there is no solution for this.

I don't think Elementor would make the official change for this as well. There are tons of websites that have been using the manual padding method instead of relying on the CSS fix.

**Comment by seriusokhatsky** [comment]
Thank you for clarification @daviedR

**Comment by black-eye** [comment]
@seriusokhatsky 
> So there is no solution to this in Elementor 3.0?

There's a legacy mode in 3.0 beta 2. But this means, you won't be able to get rid of the 2 others div, which were removed in favour of DOM reduction. 

So we can still (kind of) use the hack. 

Nevertheless this doesn't solve the main problem: the layout system used in Elementor is designed really poorly. What does bother me the most is, that Elementor developers doesn't bother to solve it at all. It's a shame, especially when they declare themselves as the most advanced design tool for WP. You can design the pixel perfect layouts, but only until you use different gaps between the columns and your layout get broken.

**Comment by seriusokhatsky** [comment]
@black-eye 

Yes, I agree with that. And I understand that it is hard to rework the layout since it may break tons of already built websites. 
But anyway, there is should be some alternative solution. Because for now, section columns gap is totally useless since you need to set left and right to 0 each time to align your columns with the container manually.

**Comment by black-eye** [comment]
@seriusokhatsky The saddest thing is that this could be the ideal opportunity to fix this problem. There are already DOM breaking changes in 3.0, so this could be one of them. But instead of fixing the issue they even break the custom workaround to repair it.

**Comment by whatsnewsisyphus** [comment]
Espexially since they are introducing a legacy mode, they could have
wrapped it in there to preserve existing sites. Its unfortunate when
developers commit to fundamental mistakes as their app scales well beyond
the original audience.

**Comment by bilalmalkoc** [comment]
It's strange that Elementor still hasn't fixed it. Other solutions are really not user friendly. @daviedR

**Comment by Needforname** [comment]
I also find this very annoying. I end up spending more time dealing with spacing / gap issues with my Elementor sites that I do designing them. Creating multi column layouts to look good in desktop view is simple enough, but everything turns to a mess in tablet and phone views. I end up having to create new layouts and mess with padding and margins. I'm surprised Elementor hasn't come up with a better way to handle spacing and responsive layout workflow. It's a shame because aside from this and a couple of other small gripes (bad icon sets for example) Elementor would be a great tool.

**Comment by nicholaszein** [comment]
Hello all!

📢 We have great news! 

### We're happy to announce that the issue you raised will be resolved with the new Flexbox Container in the near future! 🥳 

It will be released as a **stable** _(beta)_ Experiment, ready to be used by anyone!

With the new container, you'll be able to better control the positioning of your elements inside the container to make them align perfectly with the rest of the items on your website.

Please note that some of the suggestions here about adding negative margins, and other complex CSS solutions were discarded since this would prevent users from having control over their layout in many use cases. Negative margins are especially something we wanted to avoid at any cost (specifically by applying them by default) since they can cause many rendering issues, overlapping, and unexpected behaviors when used recursively. They also negatively affect other important features like motion effects, and control over spacing, not to mention requiring the addition of more DOM elements (to avoid all those issues), which is something we are aiming to avoid.

The Container implementation aims to give you finer control over your layouts. This means it's up to you to find the best way to handle spacing and positioning when you need to build more complex layouts. You have control now! 💪 

Thank you so much for your patience and passionate comments. It may not feel this way right now, but we do care about your feedback and take it very seriously!

Please feel free to check it out and update your plugin to the new version when it gets released! When it does, make sure to back up your website and enable the Flexbox Container experiment. You can then convert your templates to the new container on a page-by-page basis, or even on a section-by-section basis. Of course, some adjustments will be needed, but it will greatly improve your workflow and DOM output after optimizing your layouts.

I'll let you all know when it gets released! 😉 

Kind regards,

Nick

**Comment by newoceans** [comment]
Beste mailer,
dank je wel voor je bericht.

Mogelijk ontvang je vandaag nog niet direct een reactie van me;
ik ga proberen mijn werkdagen zo veel mogelijk te beperken tot woensdag en vrijdag.
Op deze manier hoop ik privé en zaken wat beter te kunnen scheiden ;-)

Noodgevallen zullen vanzelfsprekend wel zo snel mogelijk worden opgepakt.

Ik wens je nog een fijne dag, 
met veel energie en inspiratie; tot snel!

Met vriendelijke groet,
Marc Bijl

**Comment by black-eye** [comment]
> 📢 We have great news!
> **We're happy to announce that the issue you raised will be resolved with the new Flexbox Container in the near future! 🥳**

Well, yeah. All the issues have been resolved ... Except the one concerning the column gap while wrap enabled (as mentioned here https://github.com/elementor/elementor/issues/17066#issuecomment-1273775244) and separated gap values for rows and columns (as mentioned here: https://github.com/elementor/elementor/issues/16911#issuecomment-1273806069). 

**Hooray! everything will be resolved! ... In the future! 🥳**

**Comment by nicholaszein** [comment]
Hello @black-eye!

Thank you for takind the time to point those two issues.

I haven't yeat made it to both issues you mentioned. Please allow me some time to be able to actually reach them, they are in my list of priorities regarding gaps, but this one, since it's older, gets the answer first. There is a lot to answer, lots of issues to evaluate, and solve.

I just started my work here, and I'm going to get to them as soon as humanly possible. I hope we reach a point that I can please everyone's wishes, or at least give them an honest and deserved answer, until then I ask you to be a little more patient. We are getting there. But it won't happen overnight.

Take care.

**Comment by black-eye** [comment]
@nicholaszein Well, both of the links I referred to were your own comments you made few minutes before commenting here. ;) That's why I mentioned them here, because it happened I was following both of them.

I understand you just started to work here and I appreciate your work - finally at least someone has started to reply to all those long-forgotten issues. I hope it's the sign of things getting better. 

Please don't take my comment personally, I know, you are not the one responsible for the development. I just feel immense frustration - resolving the problem took so long and in the end, when the solution is presented to you as ready, you realize it resolves only half of the problem (and rest is yet to be addressed).

**Comment by nicholaszein** [comment]
Hello everybody!

### ✔️ This issue was resolved in Elementor **v3.8**! 🥳 

The Flexbox Container can give you now more control over how you want to align containers and covers all the concerns raised here. I know there are still concerns over separating column and row gaps with their own controls. This will be addressed in future versions, and there is already an issue dedicated to that (#18331). Please continue the discussion there. 🙏 

Feel free to check it out and update your plugin to the new version!

Cheers 🥂

**Comment by newoceans** [comment]
Beste mailer,
dank je wel voor je bericht.

Mogelijk ontvang je vandaag nog niet direct een reactie van me;
ik probeer mijn werkdagen zo veel mogelijk te beperken tot woensdag en vrijdag.
Op deze manier hoop ik privé en zaken wat beter te kunnen scheiden ;-)

Noodgevallen zullen vanzelfsprekend wel zo snel mogelijk worden opgepakt.

Ik wens je nog een fijne dag, 
met veel energie en inspiratie; tot snel!

Met vriendelijke groet,
Marc Bijl
